### DISCUSSION TITLE
.load() with populate is misleading/bugged(?)

### DISCUSSION THREAD

**Discussion by ropbastos** [mikro-orm/mikro-orm#5711] [Q&A]
In the codebase where the following code comes from (we're using MikroORM in production, thank you very much ❤️!) locationData is undefined even though it definitely is in the database. I figure what is happening is that since locationAnalysis is already loaded, the `.load({ populate: ["locationData"] })` does not go to the database to actually fetch the locationData, even though it is lazy and is being explicitly populated.

Is it expected behavior? If not, I will look into providing a reproduction and opening a bug issue 🙏. Thanks in advance.

```ts
@Entity()
@ObjectType()
export default class LocationAnalysis extends BaseEntity {
  @OneToOne({
    entity: () => Property,
    inversedBy: (property) => property.locationAnalysis,
    ref: true,
  })
  @Field(() => Property)
  property!: Ref<Property>;

  @Prop({
    type: "json",
    nullable: true,
    comment: "Location data generated in the last successful run of the location analysis",
    lazy: true, // load only when explicitly populated
    ref: true
  })
  @Field(() => LocationData, { nullable: true })
  locationData?: ScalarReference<LocationData>;
``` 

```ts
        const propertyOwner = await property.owner.loadOrFail();
        const isLocationAnalysisInitialized = property.locationAnalysis?.isInitialized();
        console.log(`Location analysis initialized: ${isLocationAnalysisInitialized}`);
        const locationAnalysis = await property.locationAnalysis?.load({
          populate: ["locationData"],
        });
        console.log("locationAnalysis.locationData: ", locationAnalysis?.locationData?.$);
        
        // Logs:
        // Location analysis loaded: true
        // locationAnalysis.locationData:  undefined
```

```
npm list mikro-orm
reos-backend@2.2.0 /Users/ropbastos/deepimmo/reos-backend
└─┬ @mikro-orm/core@6.2.9
  └── mikro-orm@6.2.9
  
 node -v
 v18.20.2
```

**Comment by ropbastos** [comment]
Just saw that there is a `refresh` option that resolves the issue. Is there any way of structuring things so the developers are remembered of the need to do this? Nothing warns us of anything here, until the runtime unexpected behavior.

**Comment by ropbastos** [comment]
Hey @B4nan, is it expected to get the `Ref<\?>` in the screenshot below? I'm still wrapping my head around those reference types. `locationAnalysis` below is a `@OneToOne({ ref: true })` property,  and when it is not present it seems to come as `null`, without the need for `.$`. However, the ScalarRef `locationData` seems to come as `Ref<?>` even though it is not present in this case in the DB. That made me run into a bug because I did `if (locationData) // Assume there is locationData to use here` in my code. Should I always use `.$` for lazy refs, even though I don't need to for non-lazy refs (it seems)?
<img width="734" alt="image" src="https://github.com/mikro-orm/mikro-orm/assets/46408745/78e9fa2d-df32-4806-abd8-1070f26e76d0">

```ts
@Prop({
    type: "json",
    nullable: true,
    comment: "Location data generated in the last successful run of the location analysis",
    lazy: true, // load only when explicitly populated
    ref: true,
  })
  @Field(() => LocationData, { nullable: true })
  locationData?: ScalarReference<LocationData>;
  ```

**Reply by ropbastos** [reply]
Thanks in advance again 🙏

**Reply by B4nan** [reply]
Yes, scalar reference wrapper is always there when the property is lazy, so you can use it to load the value. The `$` property is technically always there, but it shouldn't be defined on type level unless you populate the relation - and in such case the wrapped value should be `null` (or some actual JSON) and not `undefined`. Hard to guess what you are doing from this screenshot.

> even though I don't need to for non-lazy refs (it seems)?

Sounds like you are comparing lazy scalar properties with non lazy relation properties. Its about the "laziness". You say you ran into this with checks like `if (locationData)` but that's exactly the thing you cant do if your property is lazy - the value wont be there unless you populate it, this check is not telling you the value is falsey, it tells you it either is falsey or its not loaded at all.

**Reply by ropbastos** [reply]
Thanks for the answer!

Sorry if it is not clear to me still. Should ScalarReferences be typed as, in my example, say?
```
locationData!: ScalarReference<LocationData | null>;
```

That is because it seems to me the value will always be truthy due to the Reference wrapper, right?

I tried that, and it seems locationData.$ still is typed just as "LocationData" (missing null), even though at runtime it seems it can certainly be null.

**Reply by ropbastos** [reply]
Ok, I think I figured out my mistake. One should not pass `ref: true` in the decorator for ScalarReferences. If I remove that, then foo.someScalarReference is indeed falsy when there is no data in the database. Is that correct?

**Reply by B4nan** [reply]
Its optional feature, if you dont want the ref wrapper, dont use it. If you want it, you need ref: true. The value under $ should respect nullability, that might be a type level bug.

**Reply by ropbastos** [reply]
I see, I do want it!

I think there's a type bug then. If you're not quicker than me this time I might try my hand at a PR tonight or tomorrow :).

Can you help me define exactly what is expected to work/how's the intended use of everything so that I can work on my PR towards that goal?

Say I have a big JSON property I want to lazily load, but that property is also nullable at the database level. What is the intended use of decorators + Typescript annotations for it?

```ts
@Prop({
    type: "json",
    nullable: true,
    lazy: true,
    ref: true
  })
  bigJson!: ScalarReference<BigJson | null>;
  ```
  
  a. Is the above representative of the intended use?
  
  Then, the type of
  
  `entity.bigJson` should always be `Ref<?>`
  and `entity.bigJson.$`, with the $ only available after populate, should be `BigJson | null`
   
  b. Is that accurate?
  
  c. And about the runtime, we should expect that:
  
  When it is not present in the database:
  
  entity.bigJson --> Ref<?> // Which is truthy
  entity.bigJson.$ --> null // Available after populate
  
  And when it is present in the database:
  
  entity.bigJson --> Ref<?>
  entity.bigJson.$ --> a BigJson // Available after populate
  
  Are those the expected behaviors?
  
  Thanks again :)

**Reply by ropbastos** [reply]
For the record, what seems to not work for me is this:
```
and entity.bigJson.$, with the $ only available after populate, should be BigJson | null
```

The type comes out as BigJson only.

**Reply by ropbastos** [reply]
@B4nan Opened a PR to make the types match the runtime in that case, with a reproduction to go along: https://github.com/mikro-orm/mikro-orm/pull/5722

Let me know if it makes any sense haha 🫡

**Reply by B4nan** [reply]
yeah that all sounds correct
